File: sections/54_geometry/problems/d_interior_lattice_points/solution.cpp
#include <bits/stdc++.h>
using namespace std;

struct Point {
    long long x, y;
};

void print_int128(__int128 value) {
    if (value == 0) {
        cout << 0;
        return;
    }
    string digits;
    while (value > 0) {
        digits.push_back(char('0' + value % 10));
        value /= 10;
    }
    reverse(digits.begin(), digits.end());
    cout << digits;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    if (!(cin >> n)) return 0;
    vector<Point> polygon(n);
    for (Point& point : polygon) cin >> point.x >> point.y;

    __int128 signed_double_area = 0;
    __int128 boundary = 0;
    for (int index = 0; index < n; index++) {
        Point a = polygon[index];
        Point b = polygon[(index + 1) % n];
        signed_double_area += (__int128)a.x * b.y - (__int128)a.y * b.x;
        boundary += gcd(llabs(b.x - a.x), llabs(b.y - a.y));
    }
    __int128 double_area = signed_double_area >= 0 ? signed_double_area : -signed_double_area;
    __int128 interior = (double_area - boundary + 2) / 2;
    print_int128(interior);
    cout << '\n';
    return 0;
}

File: sections/54_geometry/problems/d_interior_lattice_points/solution.py
import math
import sys


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    n = data[0]
    polygon = [(data[index], data[index + 1]) for index in range(1, 2 * n + 1, 2)]

    signed_double_area = 0
    boundary = 0
    for index, (x1, y1) in enumerate(polygon):
        x2, y2 = polygon[(index + 1) % n]
        signed_double_area += x1 * y2 - y1 * x2
        boundary += math.gcd(abs(x2 - x1), abs(y2 - y1))
    interior = (abs(signed_double_area) - boundary + 2) // 2
    print(interior)


if __name__ == "__main__":
    main()
